volatility3/framework/layers/arm.py

```diff
@@ -0,0 +1,542 @@
+            """
+            try:
+                chunk_offset, page_size, layer_name = self._translate(offset)
+                chunk_size = min(page_size - (chunk_offset % page_size), length)
+                if not self._context.layers[layer_name].is_valid(
+                    chunk_offset, chunk_size
+                ):
+                    raise exceptions.InvalidAddressException(
+                        layer_name=layer_name, invalid_address=chunk_offset
+                    )
+            except (
+                exceptions.PagedInvalidAddressException,
+                exceptions.InvalidAddressException,
+            ) as excp:
+                """
+                Contiguous pages might not be mapped, but if we specifically ignore errors, we still want to read the n + 1 page as it might be mapped.
+                """
+                if not ignore_errors:
+                    raise
+                # We can jump more if we know where the page fault failed
+                if isinstance(excp, exceptions.PagedInvalidAddressException):
+                    mask = (1 << excp.invalid_bits) - 1
+                else:
+                    mask = (1 << (self._ttb_granule.bit_length() - 1)) - 1
+                length_diff = mask + 1 - (offset & mask)
+                length -= length_diff
+                offset += length_diff
+            else:
+                yield offset, chunk_size, chunk_offset, chunk_size, layer_name
+                length -= chunk_size
+                offset += chunk_size
+
+    def _get_virtual_addr_range(
+        self,
+    ) -> Tuple[int]:
+        """Returns the virtual address space range for the current context (user or kernel space)"""
+
+        # [2], see source/arch/arm64/include/asm/memory.h#L62
+        if self._virtual_addr_space == 0:
+            ttb_start = 0
+            ttb_size = 1 << (self._ttb_bitsize - 1)
+            ttb_end = ttb_start + (ttb_size - 1)
+        else:
+            ttb_end = 2**64 - 1
+            ttb_size = 1 << (self._ttb_bitsize - 1)
+            ttb_start = ttb_end - (ttb_size - 1)
+
+        return (ttb_start, ttb_end)
+
+    def is_valid(self, offset: int, length: int = 1) -> bool:
+        """Returns whether the address offset can be translated to a valid
+        address."""
+        try:
+            # TODO: Consider reimplementing this, since calls to mapping can call is_valid
+            return all(
+                [
+                    self._context.layers[layer].is_valid(mapped_offset)
+                    for _, _, mapped_offset, _, layer in self.mapping(offset, length)
+                ]
+            )
+        except exceptions.InvalidAddressException:
+            return False
+
+    def is_dirty(self, offset: int) -> bool:
+        """Returns whether the page at offset is marked dirty"""
+        return self._page_is_dirty(self._translate_entry(offset)[2])
+
+    @staticmethod
+    def _page_is_dirty(entry: int) -> bool:
```

Comment:
Dirty state is hardware managed only > Armv8.1-A. It is software managed by Linux (bit 55), so we can use it reliably and for all Arm versions.

However, we need to know how Windows manages it (add a class `WindowsMixIn(AArch64)` which overwrites `_page_is_dirty`), or directly use the hardware managed version for any OS, but it won't work for memory dumps  < Armv8.1-A.

`[1], see D8.4.6, page 5877` and `[1], see D8.4.6, page 5877` and https://developer.arm.com/documentation/102376/0200/Access-Flag/Dirty-state